frontend/src/lib/components/BranchButton.svelte

```diff
@@ -0,0 +1,460 @@
+<!-- BranchButton.svelte -->
+<script lang="ts">
+	import {
+		branchName,
+		documentTree,
+		currentFile,
+		editorText,
+		apiAddress,
+		allBranches
+	} from '$lib/main';
+	import { onMount } from 'svelte';
+	import { ToastType, addToast } from '$lib/toast';
+	import { cache } from '$lib/cache';
+	import type { Branch } from '$lib/types';
+
+	let showMenu = false;
+	let newBranchName: string = '';
+	let showInput = false;
+
+	/**
+	 * Fetches the list of branches from the GitHub API, categorizing them as protected or non-protected.
+	 *
+	 * This asynchronous function sends a GET request to the specified GitHub repository's API endpoint
+	 * to retrieve the branches. The function handles both successful and failed requests:
+	 * - On success, it returns a promise that resolves to an object containing two arrays:
+	 *   - `nonProtectedBranches`: A list of branches that are not protected.
+	 *   - `protectedBranches`: A list of branches that are marked as protected.
+	 * - On failure, it returns an object with empty arrays for both `nonProtectedBranches` and `protectedBranches`.
+	 *
+	 * @returns {Promise<{ nonProtectedBranches: Branch[]; protectedBranches: Branch[]; }>} A promise that resolves to an object with two properties:
+	 * - `nonProtectedBranches`: An array of non-protected branches.
+	 * - `protectedBranches`: An array of protected branches.
+	 * If the request fails, both arrays will be empty.
+	 *
+	 * @throws {Error} Will throw an error if the response from the API is unsuccessful, with details including the status code and error message from the server.
+	 *
+	 * @example
+	 * async function main() {
+	 *   try {
+	 *     const { nonProtectedBranches, protectedBranches } = await fetchExistingBranches();
+	 *   } catch (err) {
+	 *     console.error('Error fetching branches:', err);
+	 *   }
+	 * }
+	 */
+	async function fetchExistingBranches(): Promise<Branch[]> {
+		const response = await fetch(`${apiAddress}/api/branches`, {
+			method: 'GET',
+			credentials: 'include',
+			headers: {
+				'Content-Type': 'application/json'
+			}
+		});
+
+		// Check if the response is successful
+		if (!response.ok) {
+			const errorMessage = await response.json();
+			console.error('Failed to fetch branches:', errorMessage);
+			addToast({
+				message: `Error fetching branches: ${response.statusText}. ${JSON.stringify(errorMessage)}`,
+				type: ToastType.Error,
+				dismissible: true
+			});
+			return [];
+		}
+
+		const data = await response.json();
+		return (
+			data.data?.branches?.map((branch: string) => ({
+				name: branch.split(' (')[0],
+				isProtected: branch.includes('(protected)')
+			})) || []
```

Comment:
```suggestion
			})) ?? []
```
Use of the `??` operator here would be more idiomatic, because it explicitly says "if the before value is null, then use this value instead. `||` should evaluate to a boolean in every sane language.